tools/java/org/unicode/cldr/util/XMLNormalizingLoader.java

```diff
@@ -0,0 +1,686 @@
+        String fullFileName = null;
+        try {
+            fullFileName = f.getCanonicalPath();
+        } catch (IOException e) {
+            StringBuilder sb = new StringBuilder("Cannot read the file '");
+            sb.append(f);
+            throw new ICUUncheckedIOException(sb.toString(), e);
+        }
+        // use try-with-resources statement
+        try (
+            InputStream fis = new StripUTF8BOMInputStream(new FileInputStream(f));
+            InputStreamReader reader = new InputStreamReader(fis, Charset.forName("UTF-8"))
+        ) {
+            XMLSource source = new SimpleXMLSource(localeId);
+            XMLNormalizingHandler XML_HANDLER = new XMLNormalizingHandler(source, minimalDraftStatus);
+            XMLFileReader.read(fullFileName, reader, -1, true, XML_HANDLER);
+            if (XML_HANDLER.supplementalStatus == SupplementalStatus.NEVER_SET) {
+                throw new IllegalArgumentException("root of file must be either ldml or supplementalData");
+            }
+            source.setNonInheriting(XML_HANDLER.supplementalStatus == SupplementalStatus.NOT_SUPPLEMENTAL);
+            if (XML_HANDLER.overrideCount > 0) {
+                throw new IllegalArgumentException("Internal problems: either data file has duplicate path, or" +
+                    " CLDRFile.isDistinguishing() or CLDRFile.isOrdered() need updating: "
+                    + XML_HANDLER.overrideCount
+                    + "; The exact problems are printed on the console above.");
+            }
+            return source;
+        } catch (IOException e) {
+            StringBuilder sb = new StringBuilder("Cannot read the file '");
+            sb.append(f);
+            throw new ICUUncheckedIOException(sb.toString(), e);
+        }
+    }
+
+    private static class XMLNormalizingHandler implements AllHandler {
+        private DraftStatus minimalDraftStatus;
+        private static final boolean SHOW_START_END = false;
+        private int commentStackIndex;
+        private boolean justPopped = false;
+        private String lastChars = "";
+        private StringBuilder currentFullXPathSb = new StringBuilder("/");
+        private String comment = null;
+        private Map<String, String> attributeOrder;
+        private DtdData dtdData;
+        private XMLSource source;
+        private String lastActiveLeafNode;
+        private String lastLeafNode;
+        private SupplementalStatus supplementalStatus = SupplementalStatus.NEVER_SET;
+        private int maxDepth = 30; // just make deep enough to handle any CLDR file.
+        // orderedCounter, orderedString, and level logically form a single class that allows adding elements, but never removed.
+        private int[] orderedCounter = new int[maxDepth];
+        private String[] orderedString = new String[maxDepth];
+        private int level = 0;
+        private int overrideCount = 0;
+        /**
```

Comment:
Nitpick: you can make this line comment (`//`).
The types are private, so they don't need to generate javadoc.
A regular comment for the programmer reading this code is enough, and only takes 1 line, not 3.